### PR TITLE
[mds-agency] [mds-types] [mds-utils] Add TNC Mode

### PR DESCRIPTION
**Pull request by avatarneil**
## 📚 Purpose
Implements TNC state machine defined in https://github.com/lacuna-tech/mobility-data-specification/tree/add-tnc-mode

## 📦 Impacts:
- mds-agency
- mds-types
- mds-utils

## ✅ PR Checklist
- [ ] Merge 1.0 & Taxi to Develop prior to merging this PR

### CODE REVIEW

File: packages/mds-agency/tests/integration-tests.spec.ts [hunk 1]
```diff
@@ -1537,6 +1552,81 @@ describe('Tests for taxi modality', async () => {
...
+      .post(pathPrefix('/vehicles'))
+      .set('Authorization', AUTH)
+      .send(TEST_TNC)
+      .expect(201)
+      .end((err, result) => {
+        done(err)
+      })
+  })
+
+  for (const tncEvent of TNC_VEHICLE_EVENT) {
```
Comment (janedotx) [inline packages/mds-agency/tests/integration-tests.spec.ts:1571]: slick

File: packages/mds-agency/tests/integration-tests.spec.ts [hunk 2]
```diff
@@ -1537,6 +1552,81 @@ describe('Tests for taxi modality', async () => {
...
+            ? { trip_id: '1f943d59-ccc9-4d91-b6e2-0c5e771cbc6b', trip_state: vehicle_state as TRIP_STATE }
+            : {})
+        }
+        request
+          .post(pathPrefix(`/vehicles/${device_id}/event`))
+          .set('Authorization', AUTH)
+          .send(body)
+          .expect(201)
+          .end((err, result) => {
+            // test.string(result.body.status).is(EVENT_STATUS_MAP[taxiEvent])
```
Comment (janedotx) [inline packages/mds-agency/tests/integration-tests.spec.ts:1591]: Did you mean to comment this out? Looks like a legit test

File: packages/mds-types/index.ts [hunk 3]
```diff
@@ -262,6 +334,23 @@ export const TAXI_STATE_EVENT_MAP = TaxiStatusEventMap({
...
+export const TNC_STATE_EVENT_MAP = TncStatusEventMap({
+  available: [
+    'driver_cancellation',
+    'enter_jurisdiction',
+    'passenger_cancellation',
+    'provider_cancellation',
+    'service_start',
+    'trip_end'
+  ],
+  reserved: ['reservation_start', 'comms_restored', 'enter_jurisdiction'],
```
Comment (janedotx) [inline packages/mds-types/index.ts:346]: Shouldn't `passenger_cancellation` and `provider_cancellation` be in there too?

File: packages/mds-types/index.ts [hunk 4]
```diff
@@ -262,6 +334,23 @@ export const TAXI_STATE_EVENT_MAP = TaxiStatusEventMap({
...
+  non_operational: ['enter_jurisdiction', 'maintenance', 'service_end'],
+  on_trip: ['trip_start', 'enter_jurisdiction', 'comms_restored', 'trip_resume'],
+  elsewhere: ['leave_jurisdiction', 'comms_restored'],
+  unknown: ['comms_lost'],
+  stopped: ['trip_stop', 'reservation_stop']
```
Comment (janedotx) [inline packages/mds-types/index.ts:351]: Missing `trip_end` and the two cancellation events too, I think.

File: packages/mds-types/index.ts [hunk 5]
```diff
@@ -262,6 +334,23 @@ export const TAXI_STATE_EVENT_MAP = TaxiStatusEventMap({
...
+  available: [
+    'driver_cancellation',
+    'enter_jurisdiction',
+    'passenger_cancellation',
+    'provider_cancellation',
+    'service_start',
+    'trip_end'
+  ],
+  reserved: ['reservation_start', 'comms_restored', 'enter_jurisdiction'],
+  non_operational: ['enter_jurisdiction', 'maintenance', 'service_end'],
```
Comment (janedotx) [inline packages/mds-types/index.ts:347]: Missing `unspecified`

File: packages/mds-types/index.ts [hunk 6]
```diff
@@ -262,6 +334,23 @@ export const TAXI_STATE_EVENT_MAP = TaxiStatusEventMap({
...
+  on_trip: ['trip_start', 'enter_jurisdiction', 'comms_restored', 'trip_resume'],
```
Comment (janedotx) [inline packages/mds-types/index.ts:348]: Missing the cancellation events?

File: packages/mds-types/index.ts [hunk 7]
```diff
@@ -296,7 +385,7 @@ export type NullableOptional<T> = PickRequired<T> & NullableProperties<PickOptio
 export const ACCESSIBILITY_OPTIONS = ['wheelchair_accessible'] as const
 export type ACCESSIBILITY_OPTION = typeof ACCESSIBILITY_OPTIONS[number]
 
-export const MODALITIES = ['micromobility', 'taxi'] as const
+export const MODALITIES = ['micromobility', 'taxi', 'tnc'] as const
```
Comment (janedotx) [inline packages/mds-types/index.ts:388]: 🎉

File: packages/mds-types/index.ts (hunk 6 above)
Comment (avatarneil) [reply packages/mds-types/index.ts:348]: Yup yup, need to add those (probably missing for Taxi as well, they were added very late in the dev cycle)

File: packages/mds-types/index.ts (hunk 5 above)
Comment (avatarneil) [reply packages/mds-types/index.ts:347]: Good catch

File: packages/mds-types/index.ts (hunk 3 above)
Comment (avatarneil) [reply packages/mds-types/index.ts:346]: I think this is the opposite way from what you're thinking, this is basically the events that can get you _into_ this state, not out of it. `passenger_cancellation` and `provider_cancellation` get you into `available` (which is up above)

File: packages/mds-types/index.ts (hunk 6 above)
Comment (avatarneil) [reply packages/mds-types/index.ts:348]: I think my answer to this is the same as the comment above, pretty sure this is good to go

File: packages/mds-types/index.ts (hunk 4 above)
Comment (avatarneil) [reply packages/mds-types/index.ts:351]: Same as above comments

File: packages/mds-types/index.ts (hunk 3 above)
Comment (avatarneil) [reply packages/mds-types/index.ts:346]: Ah I see what you mean, I think the `TNC_EVENT_STATES_MAP` from way above is a little off, I'll patch that up

File: packages/mds-agency/tests/integration-tests.spec.ts (hunk 2 above)
Comment (avatarneil) [reply packages/mds-agency/tests/integration-tests.spec.ts:1591]: I think I'd commented it out in Taxi, not entirely sure why I did.. will try uncommenting and seeing if it breaks anything :)

File: packages/mds-types/index.ts (hunk 3 above)
Comment (avatarneil) [reply packages/mds-types/index.ts]: I ended up just whacking this data-type, it's very unintuitive to maintain with the new 'multi-trip state machine within the vehicle state machine' pattern, and wasn't used _anywhere_.

**Pull request merged**